pyccel/stdlib/ndarrays/ndarrays.c

```diff
@@ -373,3 +373,240 @@ int64_t     *numpy_to_ndarray_shape(int64_t *np_shape, int nd)
+    }
+}
+
+int64_t     numpy_sum_bool(t_ndarray arr)
+{
+    int64_t elements = 1;
+    for (int i = 0; i <arr.nd; i++)
+        elements *= arr.shape[i];
+    int64_t *tmp = (int64_t*)malloc(sizeof(int64_t) * arr.nd);
+    memset(tmp, 0, sizeof(tmp));
+    int64_t output = 0;
+    for (int32_t i = 0; i < elements; i++)
+    {
+        output += arr.nd_bool[get_index_from_array(arr, tmp)];
+        tmp[0]++;
+        for (int32_t j = 0; j < arr.nd; j++)
+            if (tmp[j] == arr.shape[j])
+            {
+                tmp[j] = 0;
+                tmp[j + 1]++;
+            }
+    }
+    free(tmp);
+    return output;
+}
+
+int64_t     numpy_sum_int8(t_ndarray arr)
+{
+    int64_t elements = 1;
+    for (int i = 0; i <arr.nd; i++)
+        elements *= arr.shape[i];
+    int64_t *tmp = (int64_t*)malloc(sizeof(int64_t) * arr.nd);
+    memset(tmp, 0, sizeof(tmp));
+    int64_t output = 0;
+    for (int32_t i = 0; i < elements; i++)
+    {
+        output += arr.nd_int8[get_index_from_array(arr, tmp)];
+        tmp[0]++;
+        for (int32_t j = 0; j < arr.nd; j++)
+            if (tmp[j] == arr.shape[j])
+            {
+                tmp[j] = 0;
+                tmp[j + 1]++;
+            }
+    }
+    free(tmp);
+    return output;
+}
+
+int64_t     numpy_sum_int16(t_ndarray arr)
+{
+    int64_t elements = 1;
+    for (int i = 0; i <arr.nd; i++)
+        elements *= arr.shape[i];
+    int64_t *tmp = (int64_t*)malloc(sizeof(int64_t) * arr.nd);
+    memset(tmp, 0, sizeof(tmp));
+    int64_t output = 0;
+    for (int32_t i = 0; i < elements; i++)
+    {
+        output += arr.nd_int16[get_index_from_array(arr, tmp)];
+        tmp[0]++;
+        for (int32_t j = 0; j < arr.nd; j++)
+            if (tmp[j] == arr.shape[j])
+            {
+                tmp[j] = 0;
+                tmp[j + 1]++;
+            }
+    }
+    free(tmp);
+    return output;
+}
+
+int64_t     numpy_sum_int32(t_ndarray arr)
+{
+    int64_t elements = 1;
+    for (int i = 0; i <arr.nd; i++)
+        elements *= arr.shape[i];
+    int64_t *tmp = (int64_t*)malloc(sizeof(int64_t) * arr.nd);
+    memset(tmp, 0, sizeof(tmp));
+    int64_t output = 0;
+    for (int32_t i = 0; i < elements; i++)
+    {
+        output += arr.nd_int32[get_index_from_array(arr, tmp)];
+        tmp[0]++;
+        for (int32_t j = 0; j < arr.nd; j++)
+            if (tmp[j] == arr.shape[j])
+            {
+                tmp[j] = 0;
+                tmp[j + 1]++;
+            }
+    }
+    free(tmp);
+    return output;
+}
+
+int64_t     numpy_sum_int64(t_ndarray arr)
+{
+    int64_t elements = 1;
+    for (int i = 0; i <arr.nd; i++)
+        elements *= arr.shape[i];
+    int64_t *tmp = (int64_t*)malloc(sizeof(int64_t) * arr.nd);
+    memset(tmp, 0, sizeof(tmp));
+    int64_t output = 0;
+    for (int32_t i = 0; i < elements; i++)
+    {
+        output += arr.nd_int64[get_index_from_array(arr, tmp)];

```

Comment:
is there a cleaner way of looping over the elements of an `nd_array`?

Reply:
Can't we just do:
```c
int64_t numpy_sum_int64(t_ndarray arr)
{
    int64_t sum = 0;
    for(int32_t idx(0); idx<arr.length; ++idx)
    {
        sum += arr.nd_int64[idx];
    }
    return sum;
}
```
?

Reply:
No, this only works for arrays that are not views.

Reply:
Ah of course. I'm completely forgetting about strides ^^

Reply:
Is `elements` equivalent to `arr.length`?

Reply:
@jalalium I have the same question

Reply:
There is a recursive option, but I'm not sure that it is faster or cleaner:
```c
int64_t     numpy_sum_int8(t_ndarray arr, int nd = -1, int offset = 0)
{
    if (nd == -1) {
        nd = arr.nd
    }
    int64_t stride = arr.strides[arr.nd-nd];
    int64_t shape = arr.shape[arr.nd-nd];
    int64_t output = 0;
    if (nd == 1) {
        for (int32_t j = 0; j < shape; j++)
        {
            ouput += arr.nd_int8[offset + j*stride];
        }
    }
    else {
        for (int32_t j = 0; j < shape; j++)
        {
            ouput += numpy_sum_int8(arr, nd-1, offset + j*stride);
        }
    }
    return output;
}
```

Reply:
Assuming the strides contain what they are supposed to (I'm not sure about this, this may be the cause of #1288). What about:

```c
int64_t     numpy_sum_int8(t_ndarray arr)
{
    int64_t local_stride[arr.nd];
    local_stride[arr.nd-1] = 1;
    for (int32_t i = arr.nd-2; i >= 0; --i) {
        local_stride[i] = local_stride[i+1]*arr.shape[i+1];
    }
    int64_t output = 0;
    for (int32_t view_index = 0; view_index < arr.length; view_index++)
    {
        int32_t i = view_index;
        int32_t array_index = 0;
        for (int32_t dim = 0; dim < arr.nd; dim++)
        {
            array_index += (i_copy / local_stride[dim]) * arr.stride[dim];
            i -= local_stride[dim] * index[dim];
        }
        output += arr.nd_int8[array_index];
    }
    return output;
}
```

Reply:
This is exactly what I am doing your `local_stride` array is my `tmp` array, I should declare it on the stack as you did. And you start with the array full and subtract from the indices while i do the opposite. I also tested the recursive method but it gave me a wrong output. I was not sure that the number of elements in a view is equivalent to `arr.length` so i will just use it and refactor the code, document it and clean it up

Reply:
I don't think so. Your `tmp` is a list of indexes which is why you can pass it to `get_index_from_array`. My solution avoids the `if` which may help

> And you start with the array full and subtract from the indices while i do the opposite. 

Not sure what you mean by "start with the array full and subtract from the indices". I iterate in order.

> I also tested the recursive method but it gave me a wrong output.

I didn't do any tests for either implementation, there may be some minor bugs

Reply:
@EmilyBourne Is this point resolved?

Reply:
I am happy with the current implementation. It is clearer with the comments. But I didn't ask the question you(@yguclu ) and @jalalium asked if there was a cleaner way of looping over the elements. I have suggested some possible alternatives. I think it is now for you two to decide if you are happy with the current implementation or if you prefer an alternative

Reply:
Ah. It is definitely supposed to be. Whether it always is may be related to PR #1299 . In any case the code now uses `arr.length` and still passes the tests